게임 결과 화면 ui 구현

tnwlsgk1002 opened the pull request:
## Related Issue

- resolved boostcampwm-2022/android11-almost-there#81

## 작업 내용 요약

- 게임 결과 화면 ui 구현
- 정산 금액 입력 다이얼로그 ui 구현
- 술값 정산 ui 구현

## Checklist

- [x] Merge 되는 브랜치가 올바른가?
- [x] Reviewers, Assignees, Labels, Projects, Milestone 설정을 했는가?
- [x] 스스로 코드 리뷰를 충분히 진행했는가?
- [x] 프로젝트의 스타일 가이드라인(컨벤션)을 준수하는가?

tnwlsgk1002 committed:
:lipstick: design: 급한 시바견 디자인 리소스 수정

- bg_sudden_shiba.xml 리소스 수정
- 해당 이미지를 사용하던 imageView 속성 수정

Issue #81

tnwlsgk1002 committed:
:lipstick: design: 결과 화면에 순위 RecyclerView 추가

- item_total_ranking 추가
- layout_character_img 수정

Issue #81

tnwlsgk1002 committed:
:lipstick: design: 결과 화면 속성값 일부 수정 및 순위표 속성 수정

- fragment RecyclerView 속성값 일부 수정
- item_real_time_ranking.xml layout 으로 감싸기
- item_total_ranking.xml 속성값 일부 수정

Issue #81

tnwlsgk1002 committed:
squash! :lipstick: design: 결과 화면에 순위 RecyclerView 추가

tnwlsgk1002 committed:
:lipstick: design: item_total_ranking 제거 및 item_ranking 으로 통일

- item_total_ranking 제거
- item_ranking 으로 통일

Issue #81

tnwlsgk1002 committed:
:lipstick: design: 결과 화면 ui 디자인 수정

- 결과 화면 정산 & 닫기 버튼 추가
- 기타 디자인 수정

Issue #81

tnwlsgk1002 committed:
:lipstick: design: 결과 화면 순위표 바인딩

- UserRanking 추가
- item_ranking에 바인딩

Issue #81

tnwlsgk1002 committed:
:lipstick: design: 결과 화면에서 버튼 클릭 시 이동 구현

- Navigation 생성
- 정산 클릭하면 CalculateFragment 로 이동
- 닫기 클릭하면 activity 종료
- toolbar 에 bind:title 추가

Issue #81

tnwlsgk1002 committed:
:lipstick: design: UserRankingAdapter 구현 및 dummy 데이터로 submitList

- UserRankingAdapter 구현
- ViewModel 에서 dummy 데이터로 submitList 실행
- 결과 화면 일부 디자인 변경

Issue #81

tnwlsgk1002 committed:
:wrench: chore: lottie 애니메이션 dependencies 추가

- lottie 5.2.0 버전 dependencies 추가

Issue #81

tnwlsgk1002 committed:
:lipstick: design: 정산 금액 입력 Dialog 추가

- dialog_amount_due 추가
- beer.json 추가
- CalculateFragment 이동 시 바로 띄워주도록 구현

Issue #81

tnwlsgk1002 committed:
:lipstick: design: 정산 금액 확인 ui 구현

- item_user_payment.xml 추가
- money.json 추가
- fragment_calculate.xml 수정
- UserPayment 추가 및 바인딩

Issue #81

tnwlsgk1002 committed:
:lipstick: design: UserPaymentAdapter 구현 및 dummy 데이터로 submitList

- UserPaymentAdapter.kt 추가
- viewModel에서 dummy 데이터로 emit 및 submitList
- 기타 디자인 수정

Issue #81

tnwlsgk1002 committed:
Merge branch 'develop' into feature/game-result-ui

tnwlsgk1002 committed:
Merge pull request #88 from tnwlsgk1002/feature/game-result-ui

게임 결과 화면 ui 구현

soulplay95 commented on presentation/src/main/res/layout/fragment_find_promise.xml:
```diff
@@ -40,13 +40,15 @@
             android:scaleType="centerCrop"
             android:src="@drawable/bg_sudden_shiba"
-            app:layout_constraintBottom_toBottomOf="@+id/container_soon_promise"
+            android:layout_marginStart="-60dp"
```
margin값이 음수인 건 처음보네요. 하나 배워갑니다!

[supports negative margins for constraints](https://developer.android.com/jetpack/androidx/releases/constraintlayout#constraintlayout-2.1.0-alpha2)

soulplay95 commented on presentation/src/main/res/layout/fragment_game_result.xml:
```diff
@@ -25,19 +26,21 @@
+            app:layout_constraintBottom_toTopOf="@id/iv_result_image"
             app:layout_constraintEnd_toEndOf="parent"
-            app:layout_constraintHorizontal_bias="0.9"
+            app:layout_constraintHorizontal_bias="0.847"
```
0.847 이라는 값은 어떻게 해서 나온 값인가요?

tnwlsgk1002 replied:
직접 조절한 값입니다.....ㅎ

soulplay95 commented on presentation/src/main/res/layout/item_real_time_ranking.xml:
```diff
@@ -1,41 +1,69 @@
+                android:layout_height="0dp"
+                app:value="100"
+                app:minValue="0"
+                app:maxValue="100"
```
min, max 값은 속성으로 지정 안할 시 default 값으로 설정되게 지정할 수도 있나요?

tnwlsgk1002 replied:
HPBar.kt에서 default 값으로 0과 100 이렇게 설정을 해두긴 했습니다!

fbghgus123 reviewed (approved):
고생하셨습니다 !

soulplay95 commented on presentation/src/main/res/values/styles.xml:
```diff
@@ -2,6 +2,20 @@
 <resources>
 
     <!--  Text Style  -->
+    <style name="Text72">
+        <item name="fontFamily">@font/sans400</item>
+        <item name="android:includeFontPadding">false</item>
+        <item name="android:textSize">72sp</item>
+    </style>
+
+    <style name="Text72.Black">
+        <item name="android:textColor">?attr/colorOnSurface</item>
+    </style>
+
+    <style name="Text72.Black.Bold">
+        <item name="android:textStyle">bold</item>
+    </style>
```
매번 다른 size의 폰트가 필요할 때마다 추가하는 것 보다 Large, Medium, Small, Caption 등으로 분류하여 사용하거나, textSize만 제외한 다른 속성만 정의해두는 등의 typography 관련 style을 통일할 필요가 있어 보입니다!

tnwlsgk1002 replied:
저도 동의합니다! 저희 Size가 여기저기 범벅으로 쓰여지고 있어서 따로 리팩토링할 때 margin 값이랑 조절해보려구요!

soulplay95 commented on presentation/src/main/kotlin/com/woory/presentation/ui/gameresult/GameResultActivity.kt:
```diff
@@ -1,9 +1,33 @@
+        }
+    }
+
+    override fun onBackPressed() {
+        super.onBackPressed()
+        overridePendingTransition(R.anim.slide_in_from_left, R.anim.slide_out_to_right)
+    }
```
모든 activity에서 back 버튼을 눌렀을 때 같은 transition을 적용하는 것처럼 보이는데, 이 부분을 `BaseActivity`로 뺄 수 있을까요?

tnwlsgk1002 replied:
오.. 좋은 생각인 것 같아요! 

근데 제가 알기로는 메인 화면에서 이동할 때 왼쪽으로 화면 이동하는 게 있고 오른쪽으로 화면 이동하는 게 있거든요.

만약 그렇게 된다면 BaseActivity에서 어디로 이동하는지 정의해주도록 강제한 다음에(abstract), 그걸 바탕으로 when 절을 써서 하는 것도 좋을 것 같아요!

soulplay95 commented on presentation/src/main/kotlin/com/woory/presentation/ui/gameresult/GameResultFragment.kt:
```diff
@@ -33,4 +40,18 @@ class GameResultFragment : BaseFragment<FragmentGameResultBinding>(R.layout.frag
+        viewLifecycleOwner.lifecycleScope.launch {
+            viewLifecycleOwner.repeatOnLifecycle(Lifecycle.State.STARTED) {
+                viewModel.userRankingList.collectLatest {
+                    (binding.rvRanking.adapter as UserRankingAdapter).submitList(it)
```
타입 캐스팅이 실패했을 때 어떤 예외 상황이 있을까요?

tnwlsgk1002 replied:
타입 캐스팅 예외 상황은 발생하지 않을 것으로 알고 고려해두진 않았는데...
듣고 보니 일리가 있긴 해서 Adapter를 변수로 보내주던지 할까 싶네요... 😇

soulplay95 commented on presentation/src/main/kotlin/com/woory/presentation/ui/gameresult/GameResultViewModel.kt:
```diff
@@ -1,10 +1,52 @@
 @HiltViewModel
 class GameResultViewModel @Inject constructor(): ViewModel() {
 
+    private val _userRankingList: MutableStateFlow<List<UserRanking>> = MutableStateFlow(emptyList())
```
`List<UserRanking>?` 으로 nullable하게 만들어서 submitList(null) 하는 방법으로 백그라운드 스레드를 만들지 않도록 하는 것도 좋아보입니다!

tnwlsgk1002 replied:
MutableStateFlow에서 null로 처리하지 않으면 백그라운드 스레드에서 돌고 있게 되나요?? (잘 몰라서 여쭙습니다!)

tnwlsgk1002 replied:
일단 수정했습니다!

soulplay95 replied:
`ListAdapter` 의 submitList 함수 내부 동작에서 background thread에서 diff를 계산 후 main thread로 Adapter.notifyItem 이벤트를 dispatch하게 되는데, 이 때 null을 넣어주게 되면 백그라운드 스레드를 만들어 execute() 하기 전에 다음 코드가 동작하여 최적화 할 수 있습니다.

```        
// fast simple remove all
if (newList == null) {
    //noinspection ConstantConditions
    int countRemoved = mList.size();
    mList = null;
    mReadOnlyList = Collections.emptyList();
    // notify last, after list is updated
    mUpdateCallback.onRemoved(0, countRemoved);
    onCurrentListChanged(previousList, commitCallback);
    return;
}

...

mConfig.getBackgroundThreadExecutor().execute(new Runnable() {
    @Override
    public void run() {
```

soulplay95 reviewed (approved):
고생하셨습니다!

swjw13 reviewed (approved):
고생하셨습니👍

tnwlsgk1002 committed:
:recycle: refactor: RecyclerView 에 들어가는 list 를 nullable 로 수정

- GameResultViewModel 의 userRankingList 를 nullable 로 설정
- GameResultViewModel 의 userPaymentList 를 nullable 로 설정

Pull Request #89

swjw13 reviewed (approved):
손이 많이가는군요 💪
